src/environment.rs

```diff
@@ -174,11 +297,113 @@ impl TestEnvironment {
     }
 
     pub fn package_address(&self, package_name: &str) -> PackageAddress {
-        *self
-            .package_addresses
+        *self.package_addresses
             .get(package_name)
             .expect(format!("Package {:?} not found", package_name).as_str())
     }
+
+    /// Creates and retrieves snapshot of the TestEnvironment
+    /// IMPORTANT: The states of the following fields are dropped:
+    /// - MenifestBuilder
+    /// - instruction_counter
+    /// - instruction_ids_by_label
+    pub fn create_snapshot(&self) -> TestEnvironmentSnapshot {
+        TestEnvironmentSnapshot::from(self)
+    }
+}
+
+/// NOTE: This should only be used for single clones,
+/// since it clones by taking a snapshot and then recovering from it.
+/// For the creation of many clones, it is advised to manually snapshot
+/// and then creating as many TestEnvironments as needed from
+/// that snapshot
+impl Clone for TestEnvironment {
+    fn clone(&self) -> Self {
+        self.create_snapshot().revive()
+    }
+}
+
+pub struct TestEnvironmentSnapshot {
+    pub test_runner_snapshot: TestRunnerSnapshot,
+
+    pub package_addresses: HashMap<String, PackageAddress>,
+    pub public_key: Secp256k1PublicKey,
+    pub account: ComponentAddress,
+    pub dapp_definition: ComponentAddress,
+
+    pub admin_badge_address: ResourceAddress,
+    pub a_address: ResourceAddress,
+    pub b_address: ResourceAddress,
+    pub x_address: ResourceAddress,
+    pub y_address: ResourceAddress,
+    pub u_address: ResourceAddress,
+    pub v_address: ResourceAddress,
+    pub j_nft_address: ResourceAddress,
+    pub k_nft_address: ResourceAddress,
+}
+
+impl TestEnvironmentSnapshot {
```

Comment:
Instead of using custom `from` and `revive` functions I think it would be better to use standard `From` traits implemented for both the `TestEnvironment` and `TestEnvironmentSnapshot`